gpu/octree/include/pcl/gpu/octree/impl/internal.hpp

```diff
@@ -47,24 +47,34 @@ namespace pcl
     namespace device
     {
 
-    template <typename T>
+    template <typename T, std::size_t n>
     struct TypeTraits;
 
-    template <>
-    struct TypeTraits<pcl::PointXYZ> {
-      struct PointXYZ {
+    template <typename T>
+    struct TypeTraits<T, 16> {
+      struct Point {
         float4 p;
       };
-      using PointType = PointXYZ;
+      using PointType = Point;
     };
 
-    template <>
-    struct TypeTraits<pcl::PointXYZRGB> {
-      struct PointXYZRGB {
+    template <typename T>
+    struct TypeTraits<T, 32> {
```

Comment:
This should be (float)4*4 + (short)4*2 = 24?

Maybe use char4 instead?

Why have you chosen float4 for XYZ? instead of float3?

Reply:
PCL uses float4 for alignment reasons even if only 3 floats are required

Reply:
Thanks for the comments, Lars and Kunal! 

The code requires that both the point-types on the device and the CPU have the same size, as expressed via the assert statement:
```c++
Static<sizeof(PointType) == sizeof(typename pcl::device::OctreeImpl<T>::PointType)>::check();
```
It seems that the core point types are either of size 16, 32, or 48 and so I created adequate GPU containers for it.

Reply:
Thanks for the comment! As you said, I think using float4 seems a reasonable thing to do because of the memory layout. 

However, I also agree that we should consider the communication time between CPU and GPU.At the moment, although all data is copied, only the XYZ part of the points are used by octree. I copied the remaining parts to satisfy the static constraint mentioned above, and also to use the simple upload facility of the DeviceArray. Maybe there is a better way?